{
 "metadata": {
  "name": "",
  "signature": "sha256:662a67676a13487b345d6f9ff9813e9debecc7ad8babe8c241b4bcf8c4c23773"
 },
 "nbformat": 3,
 "nbformat_minor": 0,
 "worksheets": [
  {
   "cells": [
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "%load_ext autoreload\n",
      "%autoreload 2\n",
      "%pylab inline\n",
      "\n",
      "import os.path as op\n",
      "base_dir = '/data/'"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "The autoreload extension is already loaded. To reload it, use:\n",
        "  %reload_ext autoreload\n",
        "Populating the interactive namespace from numpy and matplotlib\n"
       ]
      }
     ],
     "prompt_number": 8
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pyspark import SparkContext, SparkConf\n",
      "print sc\n"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "<pyspark.context.SparkContext object at 0x7fec4ea06ad0>\n"
       ]
      }
     ],
     "prompt_number": 11
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Create the input for the old goomba visualization"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "from pyspark.sql import *\n",
      "sqlContext = SQLContext(sc)\n",
      "# get the gene_id -> pubmed mapping\n",
      "gene2pubmed = (sc.textFile(op.join(base_dir, \"genbank-data/gene2pubmed\"))\n",
      "                 .filter(lambda x: x[0] != '#')\n",
      "                 .map(lambda x: x.split('\\t'))\n",
      "                 .map(lambda p: {'taxid': int(p[0]), 'geneid': int(p[1]), 'pmid': int(p[2]), 'count': 1}))\n",
      "print gene2pubmed.take(1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "[{'count': 1, 'pmid': 9873079, 'taxid': 9, 'geneid': 1246500}]\n"
       ]
      }
     ],
     "prompt_number": 24
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_neighbors = (sc.textFile(op.join(base_dir, \"genbank-data/gene_neighbors\"))\n",
      "                 .filter(lambda x: x[0] != '#')\n",
      "                 .map(lambda x: x.split('\\t'))\n",
      "                 .map(lambda p:   { 'taxid': int(p[0]), \n",
      "                                    'geneid' :int(p[1]), \n",
      "                                    'start_pos': p[4],\n",
      "                                    'end_pos': p[5],\n",
      "                                    'nucleotide_accession': p[2],\n",
      "                                    'chr': p[7],\n",
      "                                    'orientation': p[6],\n",
      "                                    'assembly': p[13]}))\n",
      "\n",
      "\n",
      "gene_info = (sc.textFile(op.join(base_dir, \"genbank-data/gene_info\"))\n",
      "             .filter(lambda x: x[0] != '#')\n",
      "             .map(lambda x: x.split('\\t'))\n",
      "             .map(lambda x: { 'taxid': int(x[0]),\n",
      "                              'geneid': int(x[1]),\n",
      "                              'description': x[8],\n",
      "                              'symbol': x[2],\n",
      "                              'name': x[11],\n",
      "                              'type_of_gene': x[9]}))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 25
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_info_keyed = gene_info.map(lambda x: ((x['taxid'], x['geneid']), x))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 26
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# filter for human genes\n",
      "human_gene_pubmed = (gene2pubmed.filter(lambda x: x['taxid'] == 9606)\n",
      "                                .map(lambda x: ((x['taxid'], x['geneid']), x)))"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 27
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def reduce_count(r1, r2):\n",
      "    '''\n",
      "    A reduce function that simply counts the number of elements in the table.\n",
      "    \n",
      "    @param r1: A Row\n",
      "    @param r2: A Row\n",
      "    @return: A new Row, equal to the first Row with a summed count.\n",
      "    '''\n",
      "    #print >>sys.stderr, \"r1:\", r1\n",
      "    r1['count'] += r2['count']\n",
      "    return r1"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 28
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# count how many references each id has\n",
      "# ((taxid, geneid), row)\n",
      "counted_human_gene_pubmed = (human_gene_pubmed.reduceByKey(reduce_count))\n",
      "counted_human_gene_pubmed.take(1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 29,
       "text": [
        "[((9606, 7748), {'count': 13, 'geneid': 7748, 'pmid': 1602151, 'taxid': 9606})]"
       ]
      }
     ],
     "prompt_number": 29
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def merge_two_dicts(x, y):\n",
      "    '''Given two dicts, merge them into a new dict as a shallow copy.'''\n",
      "    z = x.copy()\n",
      "    z.update(y)\n",
      "    return z"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 30
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# filter the gene_neighbors to those in the human GRCh38 assembly\n",
      "# ((taxid, geneid), row)\n",
      "human_neighbors = (gene_neighbors.filter(lambda x: x['assembly'].find('GRCh38') >= 0)\n",
      "                           .filter(lambda x: x['assembly'].find('Primary') >= 0)\n",
      "                           .map(lambda x: ((x['taxid'], x['geneid']), x)))\n",
      "\n",
      "human_neighbors_info = (human_neighbors.join(gene_info_keyed)\n",
      "                    .map(lambda x: (x[0], merge_two_dicts(x[1][0], x[1][1]))))\n",
      "human_neighbors_info.take(1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "ename": "Py4JJavaError",
       "evalue": "An error occurred while calling o145.partitions.\n: org.apache.hadoop.mapred.InvalidInputException: Input path does not exist: file:/data/data/gene_neighbors\n\tat org.apache.hadoop.mapred.FileInputFormat.singleThreadedListStatus(FileInputFormat.java:285)\n\tat org.apache.hadoop.mapred.FileInputFormat.listStatus(FileInputFormat.java:228)\n\tat org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:313)\n\tat org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:199)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)\n\tat scala.Option.getOrElse(Option.scala:120)\n\tat org.apache.spark.rdd.RDD.partitions(RDD.scala:237)\n\tat org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)\n\tat scala.Option.getOrElse(Option.scala:120)\n\tat org.apache.spark.rdd.RDD.partitions(RDD.scala:237)\n\tat org.apache.spark.api.java.JavaRDDLike$class.partitions(JavaRDDLike.scala:64)\n\tat org.apache.spark.api.java.AbstractJavaRDDLike.partitions(JavaRDDLike.scala:46)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:231)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:381)\n\tat py4j.Gateway.invoke(Gateway.java:259)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:209)\n\tat java.lang.Thread.run(Thread.java:745)\n",
       "output_type": "pyerr",
       "traceback": [
        "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m\n\u001b[0;31mPy4JJavaError\u001b[0m                             Traceback (most recent call last)",
        "\u001b[0;32m<ipython-input-31-92e6ab06f16a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m                            .map(lambda x: ((x['taxid'], x['geneid']), x)))\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m human_neighbors_info = (human_neighbors.join(gene_info_keyed)\n\u001b[0m\u001b[1;32m      8\u001b[0m                     .map(lambda x: (x[0], merge_two_dicts(x[1][0], x[1][1]))))\n\u001b[1;32m      9\u001b[0m \u001b[0mhuman_neighbors_info\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtake\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/pyspark/rdd.pyc\u001b[0m in \u001b[0;36mjoin\u001b[0;34m(self, other, numPartitions)\u001b[0m\n\u001b[1;32m   1609\u001b[0m         \u001b[0;34m[\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'a'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m'a'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m3\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1610\u001b[0m         \"\"\"\n\u001b[0;32m-> 1611\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mpython_join\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumPartitions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1612\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1613\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mleftOuterJoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumPartitions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mNone\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/pyspark/join.pyc\u001b[0m in \u001b[0;36mpython_join\u001b[0;34m(rdd, other, numPartitions)\u001b[0m\n\u001b[1;32m     51\u001b[0m                 \u001b[0mwbuf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     52\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mv\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mw\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mv\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvbuf\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mw\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mwbuf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 53\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0m_do_python_join\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrdd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumPartitions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     54\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     55\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/pyspark/join.pyc\u001b[0m in \u001b[0;36m_do_python_join\u001b[0;34m(rdd, other, numPartitions, dispatch)\u001b[0m\n\u001b[1;32m     39\u001b[0m     \u001b[0mvs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapValues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m     \u001b[0mws\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mother\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmapValues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mv\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 41\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mvs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munion\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mws\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupByKey\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnumPartitions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflatMapValues\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mdispatch\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__iter__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     42\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     43\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/pyspark/rdd.pyc\u001b[0m in \u001b[0;36munion\u001b[0;34m(self, other)\u001b[0m\n\u001b[1;32m    527\u001b[0m                       self.ctx.serializer)\n\u001b[1;32m    528\u001b[0m         if (self.partitioner == other.partitioner and\n\u001b[0;32m--> 529\u001b[0;31m                 self.getNumPartitions() == rdd.getNumPartitions()):\n\u001b[0m\u001b[1;32m    530\u001b[0m             \u001b[0mrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartitioner\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartitioner\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    531\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mrdd\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/pyspark/rdd.pyc\u001b[0m in \u001b[0;36mgetNumPartitions\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   2361\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2362\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mgetNumPartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2363\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_prev_jrdd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpartitions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2365\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/lib/py4j-0.9-src.zip/py4j/java_gateway.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m    811\u001b[0m         \u001b[0manswer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgateway_client\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msend_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcommand\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    812\u001b[0m         return_value = get_return_value(\n\u001b[0;32m--> 813\u001b[0;31m             answer, self.gateway_client, self.target_id, self.name)\n\u001b[0m\u001b[1;32m    814\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    815\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mtemp_arg\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtemp_args\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/pyspark/sql/utils.pyc\u001b[0m in \u001b[0;36mdeco\u001b[0;34m(*a, **kw)\u001b[0m\n\u001b[1;32m     43\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mdeco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     44\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 45\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkw\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     46\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mpy4j\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprotocol\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mPy4JJavaError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     47\u001b[0m             \u001b[0ms\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjava_exception\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoString\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
        "\u001b[0;32m/home/ubuntu/apps/spark-1.6.1-bin-hadoop2.6/python/lib/py4j-0.9-src.zip/py4j/protocol.py\u001b[0m in \u001b[0;36mget_return_value\u001b[0;34m(answer, gateway_client, target_id, name)\u001b[0m\n\u001b[1;32m    306\u001b[0m                 raise Py4JJavaError(\n\u001b[1;32m    307\u001b[0m                     \u001b[0;34m\"An error occurred while calling {0}{1}{2}.\\n\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 308\u001b[0;31m                     format(target_id, \".\", name), value)\n\u001b[0m\u001b[1;32m    309\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    310\u001b[0m                 raise Py4JError(\n",
        "\u001b[0;31mPy4JJavaError\u001b[0m: An error occurred while calling o145.partitions.\n: org.apache.hadoop.mapred.InvalidInputException: Input path does not exist: file:/data/data/gene_neighbors\n\tat org.apache.hadoop.mapred.FileInputFormat.singleThreadedListStatus(FileInputFormat.java:285)\n\tat org.apache.hadoop.mapred.FileInputFormat.listStatus(FileInputFormat.java:228)\n\tat org.apache.hadoop.mapred.FileInputFormat.getSplits(FileInputFormat.java:313)\n\tat org.apache.spark.rdd.HadoopRDD.getPartitions(HadoopRDD.scala:199)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)\n\tat scala.Option.getOrElse(Option.scala:120)\n\tat org.apache.spark.rdd.RDD.partitions(RDD.scala:237)\n\tat org.apache.spark.rdd.MapPartitionsRDD.getPartitions(MapPartitionsRDD.scala:35)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:239)\n\tat org.apache.spark.rdd.RDD$$anonfun$partitions$2.apply(RDD.scala:237)\n\tat scala.Option.getOrElse(Option.scala:120)\n\tat org.apache.spark.rdd.RDD.partitions(RDD.scala:237)\n\tat org.apache.spark.api.java.JavaRDDLike$class.partitions(JavaRDDLike.scala:64)\n\tat org.apache.spark.api.java.AbstractJavaRDDLike.partitions(JavaRDDLike.scala:46)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n\tat sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:231)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:381)\n\tat py4j.Gateway.invoke(Gateway.java:259)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:133)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:209)\n\tat java.lang.Thread.run(Thread.java:745)\n"
       ]
      }
     ],
     "prompt_number": 31
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# join (K,V) and (K,W) -> (K, (V,W)) pairs\n",
      "# map (K,(V,W)) -> (K,W)\n",
      "# join the genes with reference counts with the gene_neighbors information\n",
      "human_neighbors_info_pubmed = (counted_human_gene_pubmed.join(human_neighbors_info)\n",
      "                      .map(lambda x: ((x[1][0]['count'], x[0][0], x[0][1]), x[1][1])))\n",
      "                      #.map(lambda x: x['start_end_pos'] = (x['nucleotide_accession'], x['orientation'], x['start_pos'], x['end_pos']))\n",
      "human_neighbors_info_pubmed.take(1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "def consolidate_start_and_end(r):\n",
      "    '''\n",
      "    Consolidate the start and end rows\n",
      "    from a row.\n",
      "    \n",
      "    :param r: (key, {'start_pos': 1000, 'end_pos': 1010})\n",
      "    :return: (key, {'start_end_pos': set((1000, 1010))}\n",
      "    '''\n",
      "    r[1]['start_end_pos'] = set([(r[1]['chr'], r[1]['orientation'],int(r[1]['start_pos']), int(r[1]['end_pos']))])\n",
      "    return (r[0], r[1])\n",
      "\n",
      "def reduce_by_start_end_pos(r1,r2):\n",
      "    '''\n",
      "    Reduce all of the rows by their start / send positions.\n",
      "    \n",
      "    :param r: {'start_end_pos': set((1000, 1010))}\n",
      "    '''\n",
      "    #print >>sys.stderr, \"r1:\", r1\n",
      "    r1['start_end_pos'] = r1['start_end_pos'].union(r2['start_end_pos'])\n",
      "    return r1"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "reduced_human_neighbors_info_pubmed = (human_neighbors_info_pubmed.map(consolidate_start_and_end)\n",
      "                               .reduceByKey(reduce_by_start_end_pos))\n",
      "\n",
      "              \n",
      "reduced_human_neighbors_info_pubmed.sortByKey(ascending=False)\n",
      "reduced_human_neighbors_info_pubmed.take(1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "# take every (chr, orientation, start, end) tuple from the set and create one\n",
      "# big list out of it\n",
      "# then convert it all to TSV strings\n",
      "flattened_human_neighbors_info_pubmed = (reduced_human_neighbors_info_pubmed.flatMap(\n",
      "                                 lambda x: [[x[0][0], x[1]['geneid'], x[1]['symbol'], x[1]['name']] + list(y) for y in x[1]['start_end_pos']])\n",
      "                                 .map(lambda x: \"\\t\".join(map(str, x)) + \"\\n\")) \n",
      "all_values = flattened_human_neighbors_info_pubmed.collect()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "with open('/scr/fluidspace/pkerp/projects/goomba/output/genes_by_popularity.tsv', 'w') as f:\n",
      "    f.write('count\\tgeneid\\tsymbol\\tname\\tchr\\torientation\\tstart\\tend\\n')\n",
      "    f.writelines(all_values)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": []
    },
    {
     "cell_type": "markdown",
     "metadata": {},
     "source": [
      "## Getting top RNA genes"
     ]
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "gene_info.take(1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 15,
       "text": [
        "[{'description': u'Record to support submission of GeneRIFs for a gene not in Gene (Azorhizobium caulinodans Dreyfus et al. 1988; Azotirhizobium caulinodans.  Use when strain, subtype, isolate, etc. is unspecified, or when different from all specified ones in Gene.).',\n",
        "  'geneid': 5692769,\n",
        "  'name': u'-',\n",
        "  'symbol': u'NEWENTRY',\n",
        "  'taxid': 7,\n",
        "  'type_of_gene': u'other'}]"
       ]
      }
     ],
     "prompt_number": 15
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "rna_gene_info = (gene_info\n",
      " .filter(lambda x: x['type_of_gene'] != 'other')\n",
      " .filter(lambda x: x['type_of_gene'] != 'protein-coding')\n",
      " .filter(lambda x: x['type_of_gene'] != 'pseudo')\n",
      " .map(lambda x: ((x['taxid'], x['geneid']), x)))\n",
      "rna_gene_info.take(1)"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 16,
       "text": [
        "[((394, 7788539),\n",
        "  {'description': u'ncRNA',\n",
        "   'geneid': 7788539,\n",
        "   'name': u'-',\n",
        "   'symbol': u'NGR_b08000',\n",
        "   'taxid': 394,\n",
        "   'type_of_gene': u'ncRNA'})]"
       ]
      }
     ],
     "prompt_number": 16
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "counted_rna_gene_info = (rna_gene_info\n",
      "                         .join(counted_human_gene_pubmed)\n",
      "                         .map(lambda x: ((x[1][1]['count'], x[0][0], x[0][1]), x))\n",
      "                         .sortByKey(ascending=False)\n",
      "                         )\n",
      "nc_rnas = counted_rna_gene_info.collect()"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [],
     "prompt_number": 17
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "nc_rnas[:10]"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "metadata": {},
       "output_type": "pyout",
       "prompt_number": 18,
       "text": [
        "[((419, 9606, 406991),\n",
        "  ((9606, 406991),\n",
        "   ({'description': u'microRNA 21',\n",
        "     'geneid': 406991,\n",
        "     'name': u'microRNA 21',\n",
        "     'symbol': u'MIR21',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 419, 'geneid': 406991, 'pmid': 11679670, 'taxid': 9606}))),\n",
        " ((224, 9606, 406938),\n",
        "  ((9606, 406938),\n",
        "   ({'description': u'microRNA 146a',\n",
        "     'geneid': 406938,\n",
        "     'name': u'microRNA 146a',\n",
        "     'symbol': u'MIR146A',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 224, 'geneid': 406938, 'pmid': 12007417, 'taxid': 9606}))),\n",
        " ((220, 9606, 100189401),\n",
        "  ((9606, 100189401),\n",
        "   ({'description': u'transfer RNA-Ile (AAT) 7-2',\n",
        "     'geneid': 100189401,\n",
        "     'name': u'transfer RNA-Ile (AAT) 7-2',\n",
        "     'symbol': u'TRI-AAT7-2',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'tRNA'},\n",
        "    {'count': 220, 'geneid': 100189401, 'pmid': 17554300, 'taxid': 9606}))),\n",
        " ((216, 9606, 406947),\n",
        "  ((9606, 406947),\n",
        "   ({'description': u'microRNA 155',\n",
        "     'geneid': 406947,\n",
        "     'name': u'microRNA 155',\n",
        "     'symbol': u'MIR155',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 216, 'geneid': 406947, 'pmid': 12007417, 'taxid': 9606}))),\n",
        " ((177, 9606, 407040),\n",
        "  ((9606, 407040),\n",
        "   ({'description': u'microRNA 34a',\n",
        "     'geneid': 407040,\n",
        "     'name': u'microRNA 34a',\n",
        "     'symbol': u'MIR34A',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 177, 'geneid': 407040, 'pmid': 12554860, 'taxid': 9606}))),\n",
        " ((151, 9606, 406937),\n",
        "  ((9606, 406937),\n",
        "   ({'description': u'microRNA 145',\n",
        "     'geneid': 406937,\n",
        "     'name': u'microRNA 145',\n",
        "     'symbol': u'MIR145',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 151, 'geneid': 406937, 'pmid': 12007417, 'taxid': 9606}))),\n",
        " ((136, 9606, 7012),\n",
        "  ((9606, 7012),\n",
        "   ({'description': u'telomerase RNA component',\n",
        "     'geneid': 7012,\n",
        "     'name': u'telomerase RNA component',\n",
        "     'symbol': u'TERC',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 136, 'geneid': 7012, 'pmid': 7544491, 'taxid': 9606}))),\n",
        " ((131, 9606, 407006),\n",
        "  ((9606, 407006),\n",
        "   ({'description': u'microRNA 221',\n",
        "     'geneid': 407006,\n",
        "     'name': u'microRNA 221',\n",
        "     'symbol': u'MIR221',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 131, 'geneid': 407006, 'pmid': 12624257, 'taxid': 9606}))),\n",
        " ((109, 9606, 283120),\n",
        "  ((9606, 283120),\n",
        "   ({'description': u'H19, imprinted maternally expressed transcript (non-protein coding)',\n",
        "     'geneid': 283120,\n",
        "     'name': u'H19, imprinted maternally expressed transcript (non-protein coding)',\n",
        "     'symbol': u'H19',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 109, 'geneid': 283120, 'pmid': 1688465, 'taxid': 9606}))),\n",
        " ((104, 9606, 406952),\n",
        "  ((9606, 406952),\n",
        "   ({'description': u'microRNA 17',\n",
        "     'geneid': 406952,\n",
        "     'name': u'microRNA 17',\n",
        "     'symbol': u'MIR17',\n",
        "     'taxid': 9606,\n",
        "     'type_of_gene': u'ncRNA'},\n",
        "    {'count': 104, 'geneid': 406952, 'pmid': 11679670, 'taxid': 9606})))]"
       ]
      }
     ],
     "prompt_number": 18
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [
      "for nc in nc_rnas[:500]:\n",
      "    nc_info = nc[1][1][0]\n",
      "    print nc_info['geneid']"
     ],
     "language": "python",
     "metadata": {},
     "outputs": [
      {
       "output_type": "stream",
       "stream": "stdout",
       "text": [
        "406991\n",
        "406938\n",
        "100189401\n",
        "406947\n",
        "407040\n",
        "406937\n",
        "7012\n",
        "407006\n",
        "283120\n",
        "406952\n",
        "406906\n",
        "100048912\n",
        "406935\n",
        "407007\n",
        "406913\n",
        "407021\n",
        "406950\n",
        "406881\n",
        "406992\n",
        "407018\n",
        "406988\n",
        "406973\n",
        "406985\n",
        "4549\n",
        "407008\n",
        "4567\n",
        "407035\n",
        "406984\n",
        "406986\n",
        "406983\n",
        "7839\n",
        "574501\n",
        "406982\n",
        "406948\n",
        "100124700\n",
        "407015\n",
        "406903\n",
        "406942\n",
        "406928\n",
        "406911\n",
        "407004\n",
        "406996\n",
        "406958\n",
        "560\n",
        "406907\n",
        "10984\n",
        "494324\n",
        "407029\n",
        "406940\n",
        "407975\n",
        "406933\n",
        "407041\n",
        "406910\n",
        "406892\n",
        "50652\n",
        "55384\n",
        "407042\n",
        "407046\n",
        "406995\n",
        "406989\n",
        "407050\n",
        "406904\n",
        "406900\n",
        "406941\n",
        "406934\n",
        "406893\n",
        "7503\n",
        "574411\n",
        "406971\n",
        "378938\n",
        "406901\n",
        "407053\n",
        "407026\n",
        "407000\n",
        "406953\n",
        "406932\n",
        "407012\n",
        "407010\n",
        "406976\n",
        "406955\n",
        "406922\n",
        "125050\n",
        "5820\n",
        "407024\n",
        "407019\n",
        "406954\n",
        "406902\n",
        "407043\n",
        "407009\n",
        "406967\n",
        "406921\n",
        "50941\n",
        "10866\n",
        "442904\n",
        "407011\n",
        "406979\n",
        "406972\n",
        "406961\n",
        "406899\n",
        "574447\n",
        "442890\n",
        "407014\n",
        "406919\n",
        "6315\n",
        "574452\n",
        "407048\n",
        "406959\n",
        "6023\n",
        "4574\n",
        "102724766\n",
        "407028\n",
        "406885\n",
        "406884\n",
        "406987\n",
        "442920\n",
        "407055\n",
        "406943\n",
        "442918\n",
        "407017\n",
        "406915\n",
        "406890\n",
        "574506\n",
        "574455\n",
        "494327\n",
        "407037\n",
        "406994\n",
        "406929\n",
        "554210\n",
        "406966\n",
        "406912\n",
        "407031\n",
        "51214\n",
        "50818\n",
        "554213\n",
        "494336\n",
        "442901\n",
        "406931\n",
        "406920\n",
        "55000\n",
        "6066\n",
        "4566\n",
        "643714\n",
        "574456\n",
        "494335\n",
        "442915\n",
        "407030\n",
        "407013\n",
        "406936\n",
        "619552\n",
        "442900\n",
        "407054\n",
        "407039\n",
        "407033\n",
        "406980\n",
        "406883\n",
        "282706\n",
        "114614\n",
        "442891\n",
        "406997\n",
        "406968\n",
        "406960\n",
        "406957\n",
        "406949\n",
        "406918\n",
        "283131\n",
        "100033819\n",
        "442917\n",
        "442909\n",
        "407051\n",
        "407022\n",
        "406993\n",
        "406978\n",
        "406891\n",
        "406882\n",
        "60674\n",
        "100126355\n",
        "100126299\n",
        "693159\n",
        "574032\n",
        "442892\n",
        "406924\n",
        "145624\n",
        "8847\n",
        "4550\n",
        "724033\n",
        "652995\n",
        "574031\n",
        "442906\n",
        "407049\n",
        "407047\n",
        "407034\n",
        "407032\n",
        "407002\n",
        "406975\n",
        "406951\n",
        "406925\n",
        "406909\n",
        "51352\n",
        "10301\n",
        "1283\n",
        "101805488\n",
        "619554\n",
        "442912\n",
        "442893\n",
        "406999\n",
        "406969\n",
        "406917\n",
        "406916\n",
        "406914\n",
        "406905\n",
        "406888\n",
        "388815\n",
        "386607\n",
        "378805\n",
        "157273\n",
        "10255\n",
        "100151683\n",
        "693122\n",
        "441108\n",
        "407027\n",
        "406977\n",
        "406956\n",
        "406908\n",
        "406894\n",
        "406886\n",
        "29931\n",
        "10230\n",
        "100533107\n",
        "100359403\n",
        "100126333\n",
        "693235\n",
        "574444\n",
        "574413\n",
        "574412\n",
        "494332\n",
        "494330\n",
        "442919\n",
        "442916\n",
        "442913\n",
        "442902\n",
        "407056\n",
        "407045\n",
        "407036\n",
        "407020\n",
        "407001\n",
        "406981\n",
        "406974\n",
        "406946\n",
        "406930\n",
        "406923\n",
        "125144\n",
        "26827\n",
        "4568\n",
        "101928376\n",
        "101867536\n",
        "664617\n",
        "574453\n",
        "574448\n",
        "574443\n",
        "574434\n",
        "574408\n",
        "574034\n",
        "574033\n",
        "442903\n",
        "407044\n",
        "407025\n",
        "406998\n",
        "406965\n",
        "406963\n",
        "406944\n",
        "406895\n",
        "406889\n",
        "64002\n",
        "26871\n",
        "6060\n",
        "4565\n",
        "1867\n",
        "100131814\n",
        "693223\n",
        "693210\n",
        "693193\n",
        "693169\n",
        "693123\n",
        "574504\n",
        "574466\n",
        "494337\n",
        "494325\n",
        "442914\n",
        "442911\n",
        "442908\n",
        "442907\n",
        "442894\n",
        "440823\n",
        "407023\n",
        "406990\n",
        "406962\n",
        "406896\n",
        "406887\n",
        "338433\n",
        "26835\n",
        "10547\n",
        "3478\n",
        "2489\n",
        "101927324\n",
        "100507098\n",
        "100303453\n",
        "100126356\n",
        "100126336\n",
        "100008589\n",
        "780851\n",
        "729582\n",
        "727677\n",
        "723778\n",
        "693120\n",
        "677775\n",
        "574507\n",
        "574496\n",
        "574479\n",
        "574457\n",
        "574450\n",
        "574449\n",
        "574436\n",
        "494323\n",
        "442905\n",
        "407016\n",
        "401237\n",
        "285834\n",
        "253018\n",
        "118425\n",
        "90768\n",
        "84981\n",
        "84677\n",
        "80862\n",
        "23642\n",
        "7217\n",
        "4569\n",
        "618\n",
        "101929231\n",
        "101928165\n",
        "100885779\n",
        "100642175\n",
        "100507308\n",
        "100316868\n",
        "100302276\n",
        "100302246\n",
        "100302150\n",
        "100188856\n",
        "100144632\n",
        "100130460\n",
        "100126343\n",
        "100033822\n",
        "100033818\n",
        "100033817\n",
        "100033816\n",
        "100033815\n",
        "100033814\n",
        "100033813\n",
        "100033812\n",
        "100033811\n",
        "100033810\n",
        "100033809\n",
        "100033808\n",
        "100033807\n",
        "100033806\n",
        "100033805\n",
        "100033804\n",
        "100033803\n",
        "100033802\n",
        "100033801\n",
        "100033800\n",
        "100033799\n",
        "100033603\n",
        "100033460\n",
        "100033458\n",
        "100033456\n",
        "100033455\n",
        "100033454\n",
        "100033453\n",
        "100033451\n",
        "100033450\n",
        "100033449\n",
        "100033448\n",
        "100033447\n",
        "100033446\n",
        "100033445\n",
        "100033444\n",
        "100033443\n",
        "100033442\n",
        "100033441\n",
        "100033440\n",
        "100033437\n",
        "728655\n",
        "693190\n",
        "645323\n",
        "643911\n",
        "574511\n",
        "574480\n",
        "574474\n",
        "574451\n",
        "554202\n",
        "494326\n",
        "442910\n",
        "442897\n",
        "441951\n",
        "407003\n",
        "406964\n",
        "406945\n",
        "406927\n",
        "386593\n",
        "349114\n",
        "347689\n",
        "284260\n",
        "150967\n",
        "149830\n",
        "149775\n",
        "112597\n",
        "57291\n",
        "57061\n",
        "27004\n",
        "26781\n",
        "25859\n",
        "9808\n",
        "9383\n",
        "8123\n",
        "6043\n",
        "6029\n",
        "4572\n",
        "3378\n",
        "2221\n",
        "102724137\n",
        "101929710\n",
        "101929222\n",
        "101928694\n",
        "101752399\n",
        "101669764\n",
        "100652933\n",
        "100526833\n",
        "100507056\n",
        "100507053\n",
        "100506023\n",
        "100505832\n",
        "100423036\n",
        "100422992\n",
        "100422912\n",
        "100422893\n",
        "100302692\n",
        "100302263\n",
        "100302221\n",
        "100302201\n",
        "100302133\n",
        "100302129\n",
        "100302115\n",
        "100289178\n",
        "100271873\n",
        "100130889\n",
        "100126337\n",
        "100126334\n",
        "100126317\n",
        "100008588\n",
        "768213\n",
        "724025\n",
        "693218\n",
        "693189\n",
        "693175\n",
        "646576\n",
        "645520\n",
        "619563\n",
        "619555\n",
        "619553\n",
        "574515\n",
        "574458\n",
        "574445\n",
        "574442\n",
        "574441\n",
        "574435\n",
        "574409\n",
        "554214\n",
        "497258\n",
        "494331\n",
        "494328\n",
        "442899\n",
        "442898\n",
        "407052\n",
        "406926\n",
        "404266\n",
        "401022\n",
        "400955\n",
        "400550\n",
        "399959\n",
        "399717\n",
        "388796\n",
        "353125\n",
        "352961\n",
        "348249\n",
        "347745\n",
        "339789\n",
        "339201\n",
        "284900\n",
        "284440\n",
        "257203\n"
       ]
      }
     ],
     "prompt_number": 19
    },
    {
     "cell_type": "code",
     "collapsed": false,
     "input": [],
     "language": "python",
     "metadata": {},
     "outputs": []
    }
   ],
   "metadata": {}
  }
 ]
}